celery_serverless/worker_management.py

```diff
@@ -48,22 +49,40 @@ def spawn_worker(softlimit:'seconds'=None, hardlimit:'seconds'=None, **options):
 
 
 def shutdown_when_done(*args, **kwargs):
-    # After got a task, trigger Ctrl+C
+    # Start to shutdown by triggering <Ctrl+C>
     # The worker will stop getting tasks and will prepare to shutdown
     # but will wait the existing task to finish its lifetime.
-    logger.debug('Informing the worker to stop accepting tasks')
-    pid = os.getppid()  # Should hit my manager process. Not me, the minion process
+    logger.debug('Informing the workers to stop accepting tasks')
+    pid = os.getpid()
     os.kill(pid, signal.SIGINT)  # Trigger Ctrl+C behaviours
 
 
-def attach_hooks():
+def wakeme_soon(callback:'callable'=None, delay:'seconds'=1.0, reason='', *args, **kwargs):
+    # After born, wait some seconds and get a UNIX signal poke.
+    if reason:
+        reason = 'waiting for "%s"' % reason
+    logger.debug('Registering an ALRM for %.2f seconds ahead %s', delay, reason)
+    signal.signal(signal.SIGALRM, callback)
+    signal.setitimer(signal.ITIMER_REAL, delay)
+
+
+def attach_hooks(wait_connection=4.0, wait_job=1.0):
     """
     Register the needed hooks:
     - One to trigger stop getting tasks after the 1st, and shutdown when done
     """
-    logger.debug('Attaching hooks')
+    logger.info('Attaching hooks')
+    logger.debug('Wait connection time: %.2f', wait_connection)
+    logger.debug('Wait job time: %.2f', wait_job)
 
-    # Return the hooks to be hold somewhere. Else it could be garbage collected
-    hooks = []
-    hooks.append(task_prerun.connect(shutdown_when_done))
+    hooks = [   # Hold the hooks created. Else it could be garbage collected
+        celeryd_init.connect(partial(  # After worker process up: celeryd_init
+            wakeme_soon,
+            reason='broker connection', delay=wait_connection, callback=shutdown_when_done,
+        )),
+        worker_ready.connect(partial(  # After broker queue connected: worker_ready
+            wakeme_soon,
+            reason='job to come', delay=wait_job, callback=shutdown_when_done,
+        )),
+    ]
```

Comment:
👍

Reply:
Ou:
```python
hooks = [   # Hold the hooks created. Else it could be garbage collected
    # After worker process up: celeryd_init
    celeryd_init.connect(lambda *a, *kw: wakeme_soon(reason='broker connection', delay=wait_connection, callback=shutdown_when_done)),
    # After broker queue connected: worker_ready
    worker_ready.connect(lambda *a, *kw: wakeme_soon(reason='job to come', delay=wait_job, callback=shutdown_when_done)),
]
```
Ou:
```python
def _broker_connection_timeout(*args, **kwargs):
    return wakeme_soon(reason='broker connection', delay=wait_connection, callback=shutdown_when_done)

def _worker_ready_timeout(*args, **kwargs):
    return wakeme_soon(reason='job to come', delay=wait_job, callback=shutdown_when_done)

hooks = [   # Hold the hooks created. Else it could be garbage collected
    celeryd_init.connect(_broker_connection_timeout),  # After worker process up: celeryd_init
    worker_ready.connect(_worker_ready_timeout),  # After broker queue connected: worker_ready
]
```
Ou:
```python
@celeryd_init.connect  # After worker process up
def _broker_connection_timeout(*args, **kwargs):
    return wakeme_soon(reason='broker connection', delay=wait_connection, callback=shutdown_when_done)

@worker_ready.connect  # After broker queue connected
def _worker_ready_timeout(*args, **kwargs):
    return wakeme_soon(reason='job to come', delay=wait_job, callback=shutdown_when_done)

hooks = [_broker_connection_timeout, _worker_ready_timeout]
```